casper/src/rust/equivocation_detector.rs

```diff
@@ -0,0 +1,473 @@
+                        Self::add_equivocation_child(
+                            dag,
+                            block_store,
+                            &latest_equivocating_validator_block,
+                            equivocation_base_block_seq_num,
+                            equivocation_children,
+                        )
+                    } else {
+                        Ok(equivocation_children.clone())
+                    }
+                }
+                None => {
+                    Err(KvStoreError::KeyNotFound(
+                        "justificationBlock is missing justification pointers to equivocatingValidator even though justificationBlock isn't a part of equivocationDetectedBlockHashes for this equivocation record.".to_string()
+                    ))
+                }
+            }
+        }
+    }
+
+    fn add_equivocation_child(
+        dag: &KeyValueDagRepresentation,
+        block_store: &KeyValueBlockStore,
+        justification_block: &BlockMessage,
+        equivocation_base_block_seq_num: i64,
+        equivocation_children: &Vec<BlockMessage>,
+    ) -> Result<Vec<BlockMessage>, KvStoreError> {
+        let target_seq_num = equivocation_base_block_seq_num + 1;
+
+        match Self::find_creator_justification_ancestor_with_seq_num(
+            dag,
+            justification_block,
+            target_seq_num,
+        )? {
+            Some(equivocation_child_hash) => {
+                let equivocation_child = block_store.get_unsafe(&equivocation_child_hash);
+                let mut updated_children = equivocation_children.clone();
+                updated_children.push(equivocation_child);
+                Ok(updated_children)
+            }
+            None => {
+                Err(KvStoreError::KeyNotFound(
+                    "creator justification ancestor with lower sequence number hasn't been added to the blockDAG yet.".to_string()
+                ))
+            }
+        }
+    }
+
+    fn find_creator_justification_ancestor_with_seq_num(
+        dag: &KeyValueDagRepresentation,
+        block: &BlockMessage,
+        seq_num: i64,
+    ) -> Result<Option<BlockHash>, KvStoreError> {
+        if i64::from(block.seq_num) == seq_num {
+            return Ok(Some(block.block_hash.clone()));
+        }
+
+        let start_nodes = vec![block.block_hash.clone()];
+
+        let neighbors = |block_hash: &BlockHash| -> Vec<BlockHash> {
+            proto_util::get_creator_justification_as_list_until_goal_in_memory(
+                dag,
+                block_hash,
+                |_| false,
+            )
+            .unwrap_or_else(|_| Vec::new())
+        };
+
+        let traversal_result = dag_ops::bf_traverse(start_nodes, neighbors);
+
+        for candidate_hash in traversal_result {
+            match dag.lookup_unsafe(&candidate_hash) {
+                Ok(candidate_metadata) => {
+                    if i64::from(candidate_metadata.sequence_number) == seq_num {
```

Comment:
just question: do we need or not to check validator number as well?

```suggestion
        let target_validator = &block.sender;

        for candidate_hash in traversal_result {
            match dag.lookup_unsafe(&candidate_hash) {
                Ok(candidate_metadata) => {
                    if i64::from(candidate_metadata.sequence_number) == seq_num
                        && candidate_metadata.sender == *target_validator
```

I tried to understand the code with AI, here is a comment from AI:
```
The check only compares the sequence number, not that the block was created by the same validator you’re tracking.
You can falsely accept an ancestor that matches the seq-num but belongs to a different validator, which breaks equivocation detection.
```